### ISSUE TITLE
Vault agent with AppRole auto-auth and batch tokens renews secrets on every approle login

### ISSUE THREAD

**Issue by d-honeybadger** [hashicorp/vault#11030]
**Describe the bug**
If the AppRole used in vault agent auto-auth creates **batch** tokens upon login, these tokens cannot be renewed, and vault agent has to re-login when the batch token expires. Upon re-login, it abandons the existing consul template runner, and starts a new one. Effectively, vault agent renews secrets/certs _not_ when 2/3 of the lease duration has passed, but when the batch token expires.
Since secret renewal typically leads to reloading some application that consumes it, such behaviour means that the app is reloaded/restarted every couple hours (depending on batch token ttl).

**To Reproduce**
1. Create an approle that uses batch tokens: `vault write auth/approle/role/test_approle token_ttl=1m  
    token_type=batch policies="some-policy"`
2. Create a secret, or, in my case, pki role: `vault write pki/roles/myservice max_ttl="720h" generate_lease=true`
2. Use the above approle for auto-auth in vault agent.
3. Observe vault agent generate a new certificate every minute (instead of every 30 days, according to the cert's ttl):
```
2021-03-01T17:07:55.132-0500 [INFO]  auth.handler: lifetime watcher done channel triggered
2021-03-01T17:07:55.132-0500 [INFO]  auth.handler: authenticating
2021-03-01T17:07:55.135-0500 [INFO]  auth.handler: authentication successful, sending token to sinks
2021-03-01T17:07:55.135-0500 [INFO]  auth.handler: starting renewal process
2021-03-01T17:07:55.135-0500 [INFO]  template.server: template server received new token
2021/03/01 22:07:55.135380 [INFO] (runner) stopping
2021/03/01 22:07:55.135590 [INFO] (runner) creating new runner (dry: false, once: false)
2021/03/01 22:07:55.135720 [INFO] (runner) received finish
2021/03/01 22:07:55.135937 [INFO] (runner) creating watcher
2021-03-01T17:07:55.136-0500 [INFO]  sink.file: token written: path=/home/d-honeybadger/vault_agent_test/token
2021/03/01 22:07:55.136114 [INFO] (runner) starting
2021/03/01 22:07:55.256072 [INFO] (runner) rendered "/home/d-honeybadger/vault_agent_test/templates/myservice" => "/home/d-honeybadger/vault_agent_test/certs/myservice.serial"
2021/03/01 22:07:55.256095 [INFO] (runner) executing command "echo RESTARTING!!!" from "/home/d-honeybadger/vault_agent_test/templates/myservice" => "/home/d-honeybadger/vault_agent_test/certs/myservice.serial"
2021/03/01 22:07:55.256141 [INFO] (child) spawning: echo RESTARTING!!!
```

**Expected behavior**
Expected vault agent to re-generate certificate according to this doc: https://www.vaultproject.io/docs/agent/template#renewals-and-updating-secrets

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.6.1
* Vault CLI Version (retrieve with `vault version`): 1.6.1
* Server Operating System/Architecture: Ubuntu, Linux x86_64

Vault agent configuration file(s):

```hcl
vault {
  address = "http://127.0.0.1:8200"
}

auto_auth {
  method "approle" {
    mount_path = "auth/approle"
    config = {
      role_id_file_path = "/home/d-honeybadger/vault_agent_test/role_id"
      secret_id_file_path = "/home/d-honeybadger/vault_agent_test/secret_id"
      remove_secret_id_file_after_reading = false
    }
  }
  sink "file" {
    config = {
      path = "/home/d-honeybadger/vault_agent_test/token"
    }
  }
}

template {
  source = "/home/d-honeybadger/vault_agent_test/templates/myservice"
  destination = "/home/d-honeybadger/vault_agent_test/certs/myservice.serial"
  command = "echo RESTARTING!!!"
}
```

**Comment by AlexeyFed** [comment]
I also get this bug. When the AppRole batch token is renewed, certificates are forcefully renewed, regardless of certificate TTL. 

Another piece of functionality I'd like to point out is when Vault agent is rebooted for DR or maintenance, it regenerates all certificates/secrets. It would make more sense for certificates with valid TTLs to be stored and then restored from cache and Vault queried after reboot to ensure any cached and expired certificates are renewed. 

Environment:

Vault Server Version: 1.7.1
Vault CLI Version: 1.7.1
Server OS: CentOS-7. x86_64


exit_after_auth = false
pid_file = "./pidfile"

vault {
  address = "https://(vault_ip):8200"
}

auto_auth {
   method "approle" {
       mount_path = "auth/approle"
       config = {
           role_id_file_path = "auth/my_apps_roleID"
           secret_id_file_path = "auth/my_apps_secretID"
           remove_secret_id_file_after_reading = false
       }
   }

   sink "file" {
       config = {
           path = "auth/approleToken"
       }
   }
}

cache {
  use_auto_auth_token = true 
}

listener "tcp" {
   address = "127.0.0.1:8207"
   tls_disable = true
}

template {
  source = "templates/extract_cert.tpl"
  destination = "data/extracted_cert.cert"
}
template {
  source = "templates/extract_ca.tpl"
  destination = "data/extracted_ca.txt"
}
template {
  source = "templates/extract_private_key.tpl"
  destination = "data/private.key"
}

**Comment by kingindanord** [comment]
facing something similiar here

**Comment by ncabatoff** [comment]
Hi @d-honeybadger,

Thanks for the bug report.  I think it's safe to say that it's probably a bad idea to use batch tokens with agent auto-auth, for the reasons you point out.  I don't see any way that we could batch make tokens renewable, and it's pretty fundamental to Vault's design/security model that a token revocation revokes all leases that were created using that token.

I'd like to close this bug as everything appears to be working as intended.  Before I do so, however, I'm curious: what was the motivation for using batch tokens here?

**Comment by AlexeyFed** [comment]
Hi @ncabatoff, 

I know your question was directed to @d-honeybadger but I have the same use case to this user. From my perspective it looks like you’re  misunderstanding the issue, the issue is that vault agent is renewing the certificate when the batch token is renewed, even though the certificate had a valid TTL. The issue is that the certificate is renewed when the batch token is renewed. According to the documentation the batch token should be renewed and the certificate should remain in use until it’s TTL requires Vault agent to trigger a renewal. Not have both the batch and certificate renewed at the same time, regardless of their TTLs.

**Comment by AlexeyFed** [comment]
@ncabatoff the second part of your question, why do we need batch tokens in this type of operation? The reason is simple, this way all authentication methods remain dynamic and therefore very secure. While vault agent dynamically secures the applications secrets and certificates, it uses a static secret for auth… would make more sense if it’s auth into vault server was also dynamic.

**Comment by ncabatoff** [comment]
> when the batch token is renewed

Batch tokens cannot be renewed: https://www.vaultproject.io/docs/concepts/tokens#token-type-comparison

> The reason is simple, this way all authentication methods remain dynamic and therefore very secure. 

I'm sorry, I don't understand what you mean by "dynamic" in this context.  Could you elaborate please?

**Comment by d-honeybadger** [comment]
Hi @ncabatoff ,

Thank you for looking into this! I'd argue that discarding all templated secrets when a batch token expires works on the assumption that all secrets have leases. Correct me if I'm wrong, but with certificates, unless the `generate_lease` flag (https://www.vaultproject.io/api/secret/pki#generate_lease) is set to true, there should be no lease attached to the certificate.
For example, when I generate a certificate a certificate configured _without_ the `generate_lease` flag, I get 
![image](https://user-images.githubusercontent.com/12659852/126565190-2178ecdd-cc7d-469a-8bd9-cde108bd8ec6.png)

Therefore, as far as I understand, the certificate should not be revoked even if the batch token that was used when generating it expires.
Ideally, when vault agent goes to re-login, it would only discard the secrets that have leases, and keep the certificates until their ttl expires.

As for the motivation, we use batch tokens for services for 2 reasons: 1) reduce the load on vault when storing token state; 2) if a batch token leaks it's not as big of a deal as a service token.

**Comment by d-honeybadger** [comment]
Also, the doc on leases (https://www.vaultproject.io/docs/concepts/lease) makes it sound like leases are only applicable to service tokens 
> With every dynamic secret and service type authentication token, Vault creates a lease
Is it just a weird wording? Are leases actually applicable to batch tokens?

**Comment by AlexeyFed** [comment]
^

**Comment by ncabatoff** [comment]
> Therefore, as far as I understand, the certificate should not be revoked even if the batch token that was used when generating it expires.

You're right, I oversimplified the situation and made a bad assumption.  This part is more fixable.  Agent is restarting the template server internal routine whenever it has to re-authenticate (in this case, because it can't renew), and that's what's leading to a new certificate being generated.  I won't close this issue, let's see how viable a change it might be to allow the template server to keep running upon re-authentication.

> As for the motivation, we use batch tokens for services for 2 reasons: 1) reduce the load on vault when storing token state; 2) if a batch token leaks it's not as big of a deal as a service token.

Re (1), you're right, batch tokens are a lot cheaper in terms of storage writes, though I don't imagine you have so many agent processes that having them use service tokens for auto-auth would be a big issue.  Re (2), can you explain your reasoning?  I'm not saying you're wrong, I'm just not sure I follow.  Batch tokens can't be revoked, so it seems to me that if I had to have a token leak I'd rather it be a service token.

> Is it just weird wording? Are leases actually applicable to batch tokens?

No, you're correct.  However, I wasn't speaking about the lease associated with the token itself, but rather those that would be generated for dynamic secrets creating using the token.  But it seems I was mistaken about that too, https://www.vaultproject.io/docs/concepts/tokens#token-type-comparison points out that `Dynamic Secrets Lease Assignment` for batch tokens is `Parent (if not orphan)`.  However, I was still sort of right for the wrong reasons, as it goes on to say:

> Leases created by batch tokens are constrained to the remaining TTL of the batch tokens and, if the batch token is not an orphan, are tracked by the parent. They are revoked when the batch token's TTL expires, or when the batch token's parent is revoked (at which point the batch token is also denied access to Vault).

Thus when the batch token expires, the leases created using it will also be revoked.

**Comment by d-honeybadger** [comment]
> I wasn't speaking about the lease associated with the token itself, but rather those that would be generated for dynamic secrets creating using the token

Thank you! That distinction helped a lot in understanding the whole lease + batch tokens situation.

> I won't close this issue, let's see how viable a change it might be to allow the template server to keep running upon re-authentication.

Thanks! Yeah, I was once looking at the related code and it seemed like there would need to be a change in the consul template itself, for it to be able to dynamically set a token.

> Re (1), you're right, batch tokens are a lot cheaper in terms of storage writes, though I don't imagine you have so many agent processes that having them use service tokens for auto-auth would be a big issue. Re (2), can you explain your reasoning?

Yeah, regarding the load, it's not a big deal if a few things that need vault agent processes use service tokens. As for security, all our batch tokens are short-lived, which is why IMO they're better. In order to revoke a service token, one needs to detect that it was leaked first, which, unless the "attacker" start doing something crazy with that token, isn't that simple. Whereas a batch token will automatically become useless in an hour or so.

**Comment by ncabatoff** [comment]
> As for security, all our batch tokens are short-lived, which is why IMO they're better. In order to revoke a service token, one needs to detect that it was leaked first, which, unless the "attacker" start doing something crazy with that token, isn't that simple. Whereas a batch token will automatically become useless in an hour or so.

All true.  Of course you could also have short-lived service tokens.  And note that renewal is lighter-weight than actually generating or revoking a token.

**Comment by ncabatoff** [comment]
To save someone taking this on from reading all the above, the enhancement proposal is:

Agent is restarting the template server internal routine whenever it has to re-authenticate.  Can we change it to allow the template server to keep running upon re-authentication?

**Comment by jonwilliams84** [comment]
Do we have any updates to this enhancement - we are also experiencing this issue and as it stands it makes vault-agent unfit for purpose (as our app cannot be restarted every 2 minutes!)

For the time being I may need to resort to a cron'd vault write to obtain certs, keys and CA...which is less than ideal.

**Comment by calvn** [comment]
> Another piece of functionality I'd like to point out is when Vault agent is rebooted for DR or maintenance, it regenerates all certificates/secrets. It would make more sense for certificates with valid TTLs to be stored and then restored from cache and Vault queried after reboot to ensure any cached and expired certificates are renewed.

We released a feature on Agent not too long ago called [persistent cache](https://www.vaultproject.io/docs/agent/caching/persistent-caches/kubernetes) that satisfies this exact workflow on Kubernetes for transferring state between Agent init and sidecar containers. We're looking into extending this capability to work beyond Kubernetes, but we don't have a precise timeline to share at the moment. This is something that we've definitely given some thought on though, so stay tuned!

**Comment by AlexeyFed** [comment]
> > Another piece of functionality I'd like to point out is when Vault agent is rebooted for DR or maintenance, it regenerates all certificates/secrets. It would make more sense for certificates with valid TTLs to be stored and then restored from cache and Vault queried after reboot to ensure any cached and expired certificates are renewed.
> 
> We released a feature on Agent not too long ago called [persistent cache](https://www.vaultproject.io/docs/agent/caching/persistent-caches/kubernetes) that satisfies this exact workflow on Kubernetes for transferring state between Agent init and sidecar containers. We're looking into extending this capability to work beyond Kubernetes, but we don't have a precise timeline to share at the moment. This is something that we've definitely given some thought on though, so stay tuned!

Yes, thanks for the link, however that’s the first thing that comes up when you look for that feature. I don’t use Kubernetes as I don’t want to become another Kubernetes horror story, would be good if this worked with Vault natively and Nomad as well.

**Comment by calvn** [comment]
That's totally reasonable, and we'd love to be able to support this for workflows outside of Kubernetes as well. We focused on Kubernetes upon initial release of the feature given the constraints and reported use case from our customers back then, but this thread certainly helps with raising awareness on bringing general support for persistent caching.

**Comment by aiman-alsari** [comment]
I can reproduce this issue with a short lived service token and also with caching enabled. It's not confined to just batch tokens as far as I can tell.

**Comment by Zmadell** [comment]
We've replicated this using Azure Auto Auth. Makes it unusable for our use case since the cert change triggers a reload of apache service.

**Comment by AlexeyFed** [comment]
Thanks for opening a tick, however I’m not sure why this one is labeled an enhancement when the documentation says this should work. To quote d-honeybadger:


*Expected behavior*
Expected vault agent to re-generate certificate according to this doc: https://www.vaultproject.io/docs/agent/template#renewals-and-updating-secrets

**Comment by hsimon-hashicorp** [comment]
> Thanks for opening this as a ticket, however I’m not sure why this one is labeled an enhancement when the documentation says this should work. (Maybe an update to the docs is needed) To quote d-honeybadger:
> 
> **Expected behaviour** Expected vault agent to re-generate certificate according to this doc: https://www.vaultproject.io/docs/agent/template#renewals-and-updating-secrets

Hi there - Nick mentioned why it's being labeled as an enhancement [here](https://github.com/hashicorp/vault/issues/11030#issuecomment-884561926). :)

**Comment by HelenLorenzo** [comment]
Hello,
I'm in the same situation. I need to use kubernetes authentication where the role is configured with token type batch so that the authentication works for both the main vault and the PR.
I am not able to have the batch token re-created automatically.
Do you have any suggestions?

**Comment by cleanet** [comment]
Hello. I have the same problem.
I have my vault agent running in podman. This agent generates templates for certificates.
I added the `sink` (type `file`) inside `auto_auth` and the problem of that each reboot of the service this created news certificates were solved adding the `sink`. Add my configuration:
```
#pid_file = "./pidfile"
#log_level = "debug"
vault {
  address = "https://vault.apps.k8s.losheredia.net"
  tls_skip_verify = true
  retry {
    num_retries = -1
  }
}

auto_auth {
  method {
    type = "approle"
    config = {
      role_id_file_path = "/home/vault/config/roleid"
    }
  }
  sinks {
    sink {
      type = "file"

      config = {
        path = "/home/vault/config/cachefile"
      }
    }
  }
}

template_config {
  exit_on_retry_failure = true
}

template {
  source = "/home/vault/templates/dns.losheredia.net.ctmpl"
  destination = "/certs/dns.losheredia.net/dns.losheredia.net.pem"
  error_on_missing_key = true
}
template {
  source = "/home/vault/templates/nginx.losheredia.net.ctmpl"
  destination = "/certs/nginx.losheredia.net/nginx.losheredia.net.pem"
  error_on_missing_key = true
}
template {
  source = "/home/vault/templates/gitlab.losheredia.net.ctmpl"
  destination = "/certs/gitlab.losheredia.net/gitlab.losheredia.net.pem"
  error_on_missing_key = true
}
```
FileSystem:
```
drwxr-xr-x. 2 podman podman  55 Mar 15 16:55 .
drwxr-xr-x. 5 podman podman  50 Mar  9 15:14 ..
-rw-r-----. 1 podman podman  95 Mar  9 15:22 cachefile
-rw-r--r--. 1 podman podman 995 Mar 15 16:55 config.hcl
-rw-r--r--. 1 podman podman  37 Feb 22 18:44 roleid
```

Template gitlab.losheredia.net.ctmpl:
```
{{ with pkiCert "pki_int/issue/losheredia-net" "common_name=gitlab.losheredia.net" }}{{ .CA }}{{ .Cert }}{{ .Key | trimSpace }}{{ end }}
```

Adding the `sink` works fine for me.

Thanks

Grettings.
